File: src/u2ctl/selectors/resolver.py
"""Selector priority resolution and ambiguity warning logic."""

import re
from typing import List, Dict, Any, Tuple, Optional
from u2ctl.errors import SelectorNotFoundError
from u2ctl.models import ActionElement


def parse_bounds_rect(bounds_str: str) -> Optional[Tuple[int, int, int, int]]:
    m = re.match(r"\[(\d+),(\d+)\]\[(\d+),(\d+)\]", bounds_str)
    if m:
        return (int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4)))
    return None


def rect_overlap_ratio(rect1: Tuple[int, int, int, int], rect2: Tuple[int, int, int, int]) -> float:
    """Calculate overlap ratio relative to smaller rectangle area."""
    x1 = max(rect1[0], rect2[0])
    y1 = max(rect1[1], rect2[1])
    x2 = min(rect1[2], rect2[2])
    y2 = min(rect1[3], rect2[3])

    if x2 <= x1 or y2 <= y1:
        return 0.0

    intersection = (x2 - x1) * (y2 - y1)
    area1 = (rect1[2] - rect1[0]) * (rect1[3] - rect1[1])
    area2 = (rect2[2] - rect2[0]) * (rect2[3] - rect2[1])
    smaller_area = min(area1, area2)

    if smaller_area <= 0:
        return 0.0
    return intersection / float(smaller_area)


def resolve_selector(
    elements: List[ActionElement],
    selector: Dict[str, Any],
    strict_selector: bool = False,
) -> Tuple[ActionElement, List[str]]:
    """Resolve element matching selector according to priority and ambiguity rules."""
    warnings: List[str] = []
    matched: List[ActionElement] = []

    # Priority matching: resource_id > description > text > desc_contains > text_contains > bounds
    if "resource_id" in selector:
        rid = selector["resource_id"]
        matched = [e for e in elements if e.resource_id == rid or e.resource_id.endswith(f":id/{rid}")]
    elif "description" in selector:
        desc = selector["description"]
        matched = [e for e in elements if e.content_desc == desc]
    elif "text" in selector:
        txt = selector["text"]
        matched = [e for e in elements if e.text == txt]
    elif "desc_contains" in selector:
        sub = selector["desc_contains"].lower()
        matched = [e for e in elements if sub in e.content_desc.lower()]
    elif "text_contains" in selector:
        sub = selector["text_contains"].lower()
        matched = [e for e in elements if sub in e.text.lower()]
    elif "bounds" in selector:
        req_b = selector["bounds"]  # [x1, y1, x2, y2]
        matched = []
        for e in elements:
            b = parse_bounds_rect(e.bounds)
            if b and rect_overlap_ratio((req_b[0], req_b[1], req_b[2], req_b[3]), b) >= 0.9:
                matched.append(e)

        def _get_area(elem: ActionElement) -> int:
            b = parse_bounds_rect(elem.bounds)
            if not b:
                return 2**62
            return (b[2] - b[0]) * (b[3] - b[1])

        matched.sort(key=_get_area)

    if not matched:
        raise SelectorNotFoundError(
            f"No element found matching selector {selector}",
            hint="Run `u2ctl ui dump --filter actionable` to inspect current on-screen elements",
        )

    if len(matched) == 1:
        return matched[0], warnings

    # Ambiguity resolution rules (G4)
    # 1. Focused
    focused_matches = [e for e in matched if e.focused]
    if len(focused_matches) == 1:
        return focused_matches[0], warnings

    # 2. Same screen rect
    first_b = parse_bounds_rect(matched[0].bounds)
    if first_b:
        all_same = True
        for m in matched[1:]:
            mb = parse_bounds_rect(m.bounds)
            if not mb or rect_overlap_ratio(first_b, mb) < 0.9:
                all_same = False
                break
        if all_same:
            return matched[0], warnings

    # 3. Multiple matches warning or strict failure
    if strict_selector:
        raise SelectorNotFoundError(
            f"Selector matched {len(matched)} elements in strict mode: {[m.bounds for m in matched]}",
            hint="Pass --bounds or specify exact resource-id/description to uniquely target element",
        )

    bounds_list = ", ".join(m.bounds for m in matched[:3])
    warnings.append(f"SELECTOR_MATCHED_MULTIPLE matched {len(matched)}, used first ({matched[0].bounds}). Pass --bounds to disambiguate: [{bounds_list}]")
    return matched[0], warnings

File: u2bun/src/selectors/resolver.ts
import { SelectorNotFoundError, UsageError } from "../errors";
import type { ActionElement } from "../models";
import { parseBoundsRect, type SelectorQuery } from "./parser";

export interface ResolvedElementMatch {
  element: ActionElement;
  centerX: number;
  centerY: number;
  matchedCount: number;
  warnings: string[];
}

export const OVERLAP_MERGE = 0.85;       // dedup: elements with same-size bounds collapse to one
export const OVERLAP_MATCH = 0.80;       // selector --bounds: permissive fuzzy match
export const OVERLAP_AMBIGUOUS = 0.90;   // disambiguation: all candidates overlap -> pick first

export function rectOverlapRatio(
  r1: { x1: number; y1: number; x2: number; y2: number },
  r2: { x1: number; y1: number; x2: number; y2: number }
): number {
  const ix1 = Math.max(r1.x1, r2.x1);
  const iy1 = Math.max(r1.y1, r2.y1);
  const ix2 = Math.min(r1.x2, r2.x2);
  const iy2 = Math.min(r1.y2, r2.y2);

  if (ix2 <= ix1 || iy2 <= iy1) return 0;

  const intersectionArea = (ix2 - ix1) * (iy2 - iy1);
  const area1 = (r1.x2 - r1.x1) * (r1.y2 - r1.y1);
  const area2 = (r2.x2 - r2.x1) * (r2.y2 - r2.y1);
  const minArea = Math.min(area1, area2);

  if (minArea <= 0) return 0;
  return intersectionArea / minArea;
}

function filterElementsMatchingQuery(elements: ActionElement[], query: SelectorQuery): ActionElement[] {
  return elements.filter((e) => {
    if (query.ref) {
      const matchRef = e.ref === query.ref || `@${e.index + 1}` === query.ref || `${e.index + 1}` === query.ref;
      if (!matchRef) return false;
    }
    if (query.resourceId && e.resourceId !== query.resourceId) return false;
    if (query.contentDesc && e.contentDesc !== query.contentDesc) return false;
    if (query.descContains && !e.contentDesc.toLowerCase().includes(query.descContains.toLowerCase())) return false;
    if (query.text && e.text !== query.text) return false;
    if (query.textContains && !e.text.toLowerCase().includes(query.textContains.toLowerCase())) return false;
    if (query.className && e.className !== query.className) return false;

    if (query.bounds) {
      const eRect = parseBoundsRect(e.bounds);
      if (!eRect) return false;
      const overlap = rectOverlapRatio(query.bounds, eRect);
      if (overlap < OVERLAP_MATCH) return false;
    }

    return true;
  });
}

export function resolveSelector(
  elements: ActionElement[],
  query: SelectorQuery,
  strictSelector: boolean = false,
  rawElements?: ActionElement[]
): ResolvedElementMatch {
  const warnings: string[] = [];

  let matches = filterElementsMatchingQuery(elements, query);

  if (matches.length === 0 && rawElements && rawElements.length > 0) {
    const rawMatches = filterElementsMatchingQuery(rawElements, query);
    if (rawMatches.length > 0) {
      warnings.push("Selector matched element only in raw element tree (hidden by deduplication).");
      matches = rawMatches;
    }
  }

  if (matches.length === 0) {
    const desc = JSON.stringify(query);
    throw new SelectorNotFoundError(desc);
  }

  if (matches.length === 1) {
    const target = matches[0];
    const rect = parseBoundsRect(target.bounds) || { x1: 0, y1: 0, x2: 0, y2: 0 };
    return {
      element: target,
      centerX: Math.round((rect.x1 + rect.x2) / 2),
      centerY: Math.round((rect.y1 + rect.y2) / 2),
      matchedCount: 1,
      warnings,
    };
  }

  // Ambiguity Policy (BUILDSPEC G4)
  if (strictSelector) {
    throw new SelectorNotFoundError(
      `Ambiguous selector query ${JSON.stringify(query)} matched ${matches.length} elements under --strict-selector`
    );
  }

  // Rule 1: Focused element wins
  const focusedMatches = matches.filter((m) => m.focused);
  if (focusedMatches.length === 1) {
    const target = focusedMatches[0];
    const rect = parseBoundsRect(target.bounds) || { x1: 0, y1: 0, x2: 0, y2: 0 };
    return {
      element: target,
      centerX: Math.round((rect.x1 + rect.x2) / 2),
      centerY: Math.round((rect.y1 + rect.y2) / 2),
      matchedCount: matches.length,
      warnings,
    };
  }

  // Rule 2: Overlapping rects (>=90%)
  const firstRect = parseBoundsRect(matches[0].bounds);
  if (firstRect) {
    const allOverlap = matches.every((m) => {
      const r = parseBoundsRect(m.bounds);
      return r && rectOverlapRatio(firstRect, r) >= OVERLAP_AMBIGUOUS;
    });
    if (allOverlap) {
      const target = matches[0];
      return {
        element: target,
        centerX: Math.round((firstRect.x1 + firstRect.x2) / 2),
        centerY: Math.round((firstRect.y1 + firstRect.y2) / 2),
        matchedCount: matches.length,
        warnings,
      };
    }
  }

  // Rule 3: First in document order + Warning
  const selected = matches[0];
  const boundsList = matches.map((m) => m.bounds).join(", ");
  warnings.push(`Ambiguous selector matched ${matches.length} elements (${boundsList}). Selected first element in document order.`);

  const rect = parseBoundsRect(selected.bounds) || { x1: 0, y1: 0, x2: 0, y2: 0 };
  return {
    element: selected,
    centerX: Math.round((rect.x1 + rect.x2) / 2),
    centerY: Math.round((rect.y1 + rect.y2) / 2),
    matchedCount: matches.length,
    warnings,
  };
}
